src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -38,89 +41,505 @@ public static ControlFlowGraph Create(IBlockOperation body)
+                    {
+                        Debug.Assert(currentRegion.FirstBlock == block);
+                        continue;
+                    }
+
+                    if (next.Destination == null)
+                    {
+                        if ((next.Flags & BasicBlock.BranchFlags.StructuredExceptionHandling) == 0)
+                        {
+                            continue;
+                        }
+
+                        ImmutableHashSet<BasicBlock> predecessors = block.Predecessors;
+
+                        // PROTOTYPE(dataflow): It should be safe to merge other branches with null destination, even when there are more than one predecessor 
+                        //                      and more than one incoming branch
+
+                        if (predecessors.Count != 1)
+                        {
+                            continue;
+                        }
+
+                        BasicBlock predecessor = predecessors.Single();
+
+                        if (predecessor.Ordinal != i - 1 ||
+                            predecessor.InternalNext.Destination != block ||
+                            predecessor.InternalConditional.Branch.Destination == block ||
+                            regionMap[predecessor] != currentRegion)
+                        {
+                            // Do not merge StructuredExceptionHandling into the middle of the filter or finally,
+                            // Do not merge StructuredExceptionHandling into conditional branch
+                            // Do not merge StructuredExceptionHandling into a different region
+                            continue;
+                        }
+
+                        predecessor.InternalNext = next;
+                    }
+                    else
+                    {
+                        Debug.Assert((next.Flags & ~(BasicBlock.BranchFlags.Regular | BasicBlock.BranchFlags.Error)) == 0);
+
+                        RegionBuilder destinationRegion = regionMap[next.Destination];
+
+                        // If source and destination are in different regions, it might
+                        // be unsafe to merge branches.
+                        if (currentRegion != destinationRegion)
                         {
-                            if (predecessor.Next == block)
+                            fromCurrent?.Clear();
+                            fromDestination?.Clear();
+
+                            if (!checkBranchesFromPredecessors())
                             {
-                                predecessor.Next = next;
-                                next.AddPredecessor(predecessor);
+                                continue;
                             }
 
-                            (IOperation condition, bool jumpIfTrue, BasicBlock destination) = predecessor.Conditional;
-                            if (destination == block)
+                            bool checkBranchesFromPredecessors()
```

Comment:
Overall, I find the `PackBlocks` pretty complicated too. It's long, nested deeply and has local functions interspersed with main logic flow.
Could `checkBranchesFromPredecessors`, `collectAncestorsAndSelf` and `getIndexOfLastLeftRegion` be moved to the end of the method? I found that helped for the verifier logic. 
I'd also consider breaking out into multiple methods if some blocks are suitable.

Reply:
Thanks. That helped!

---
In reply to: [167966182](https://github.com/dotnet/roslyn/pull/24770#discussion_r167966182) [](ancestors = 167966182)